exonum/src/runtime/rust/call_context.rs

```diff
@@ -0,0 +1,200 @@
+            instance,
+        }
+    }
+
+    /// Returns a writable snapshot of the current blockchain state.
+    pub fn fork(&self) -> &Fork {
+        self.inner.fork
+    }
+
+    /// Returns the initiator of the actual transaction execution.
+    pub fn caller(&self) -> &Caller {
+        &self.inner.caller
+    }
+
+    /// Accesses dispatcher information.
+    pub fn dispatcher(&self) -> Schema<&Fork> {
+        Schema::new(self.fork())
+    }
+
+    pub fn instance(&self) -> InstanceDescriptor {
+        self.instance
+    }
+
+    #[doc(hidden)]
+    pub fn call(
+        &mut self,
+        interface_name: impl AsRef<str>,
+        method_id: MethodId,
+        arguments: impl BinaryValue,
+    ) -> Result<(), ExecutionError> {
+        let call_info = CallInfo {
+            instance_id: self.instance.id,
+            method_id,
+        };
+        self.inner.call(
+            interface_name.as_ref(),
+            &call_info,
+            arguments.into_bytes().as_ref(),
+        )
+    }
+
+    // TODO This method is hidden until it is fully tested in next releases. [ECR-3493]
+    /// Creates a client to call interface methods of the specified service instance.
+    #[doc(hidden)]
+    pub fn interface<'s, T>(&'s mut self, called: InstanceId) -> Result<T, ExecutionError>
+    where
+        T: From<CallContext<'s>>,
+    {
+        self.call_context(called).map(Into::into)
+    }
+
+    // TODO This method is hidden until it is fully tested in next releases. [ECR-3493]
+    #[doc(hidden)]
+    pub fn call_context<'s>(
+        &'s mut self,
+        called_id: impl Into<InstanceQuery<'s>>,
+    ) -> Result<CallContext, ExecutionError> {
+        let descriptor = self
+            .inner
+            .dispatcher
+            .get_service(self.fork(), called_id)
+            .ok_or(DispatcherError::IncorrectInstanceId)?;
+        Ok(CallContext {
+            inner: self.inner.child_context(self.instance.id),
+            instance: descriptor,
+        })
+    }
+
+    /// Checks the caller of this method with the specified closure.
+    ///
+    /// If the closure returns `Some(value)`, then the method returns `Some((value, fork))` thus you
+    /// get a write access to the blockchain state. Otherwise this method returns
+    /// an occurred error.
+    pub fn verify_caller<F, T>(&self, predicate: F) -> Option<(T, &Fork)>
+    where
+        F: Fn(&Caller) -> Option<T>,
+    {
+        // TODO Think about returning structure with the named fields instead of unnamed tuple
+        // to make code more clear. [ECR-3222]
+        predicate(&self.inner.caller).map(|result| (result, &*self.inner.fork))
+    }
+
+    /// Isolates execution of the provided closure.
+    ///
+    /// This method should be used with extreme care, since it subverts the usual rules
+    /// of transaction roll-back provided by Exonum. Namely:
+    ///
+    /// - If the execution of the closure is successful, all changes to the blockchain state
```

Comment:
This feature is also known as multi-layered (fork) checkpoints. We've discussed it previously (since it makes sense for inter-service communication), but no implementation thus far has appeared 🤷‍♂